Floor updates

0xNom opened the pull request.

0xNom committed:
Floor updates

0xNom committed:
Update claiming tests

0xNom committed:
Remove types exclusion

0xCaps committed:
Updated tokenweth calculator, treasury and test

0xCaps committed:
Merge pull request #3 from FloorDAO/feature/claiming-tests

0xCaps committed:
Merge pull request #4 from FloorDAO/fixup/tokenweth-calculator

0xCaps committed:
fix: gFloor payout emitted incorrectly

0xaeto commented on contracts/Treasury.sol:
```diff
@@ -153,6 +161,38 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
     }
 
+    /**
+     * @notice allow our allocator to withdraw without affecting our reserves or FLOOR
```
```suggestion
     * @notice allow our allocator to deposit without affecting our reserves or FLOOR
```

0xaeto commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,285 @@
+
+    /**
+     * @notice withdraws asset from treasury, deposits asset into lending pool,
+     * then deposits xToken into treasury.
```
this step appears to be missing — depositing xToken into treasury

but it might be easier to just let the xToken sit in the allocator to save the extra step of shuffling it from allocator <> treasury

0xaeto reviewed (commented):
only a couple of minor things

0xCaps committed:
Update contracts/Treasury.sol

Co-authored-by: 0xaeto <83855517+0xaeto@users.noreply.github.com>

0xCaps committed:
Update NFTXAllocator.sol

Fixed accuracy of documentation

0xCaps committed:
Allow the treasury to hold and value xTokens:
- allocator deposits and withdrawals removed, all deposits/withdrawals are now tracked in 	otalReserves
- introduction of STATUS.XTOKEN in the treasury, which triggers a calculation in the Treasury's 	okenValue()
- xToken and xTokenWeth calculators use NFTX contracts to determine value of underlying vToken or vTokenWeth balances
- updates to interfaces and mocks

0xNom committed:
Rename calculators

0xNom committed:
Generalise comment

0xNom committed:
Update flow

0xCaps committed:
allocatorManage introduced to skip excessReserves check

0xCaps committed:
Update TreasuryMock.sol

0xCaps committed:
Merge pull request #5 from FloorDAO/feat/treasury-xtokens

Core contracts approved, remaining test fixes will be applied in a future PR.

0xCaps committed:
Update NFTXAllocator.sol

0xNom committed:
Fix allocator test

0xNom committed:
Un-skip treasury tests

0xCaps commented on contracts/Treasury.sol:
```diff
@@ -59,18 +64,19 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
     mapping(STATUS => mapping(address => bool)) public permissions;
-    mapping(address => address) public bondCalculator;
+    mapping(address => address) public override bondCalculator;
+    mapping(address => uint256) public _riskOffValuation; // 18 decimal in ETH terms 
```
9 decimal in FLOOR terms i.e. PUNK would be `20_000_000_000_000` (20,000 FLOOR or 20 ETH)

0xCaps commented on contracts/Treasury.sol:
```diff
@@ -456,21 +502,61 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+            return _amount.mul(riskOffValuation(_token)).div(10**IERC20Metadata(address(_token)).decimals());
+        }
+
+        // If our token is an XTOKEN used in NFTX then we will utilise our bonding calculator
```
Remove specific mention of NFTX

0xCaps committed:
testnet token faucet for Rinkeby

0xCaps committed:
Fix to NFTX timelock on xToken mints

0xNom committed:
Add test suite

0xCaps committed:
reduce drip amounts

0xCaps committed:
Merge pull request #6 from FloorDAO/feat/rinkeby-drip-contract

0xCaps committed:
Merge pull request #8 from FloorDAO/fixup/nftx-timelock

0xCaps committed:
Pull assets from Treasury in harvestAll

JExtor commented on contracts/types/NoteKeeper.sol:
```diff
@@ -87,37 +93,28 @@ abstract contract NoteKeeper is INoteKeeper, FrontEndRewarder {
+   * @return payout_     sum of payout sent, in gFLOOR
    */
-  function redeem(address _user, uint256[] memory _indexes, bool _sendgOHM) public override returns (uint256 payout_) {
+  function redeem(address _user, uint256[] memory _indexes, bool _sendgFLOOR) public override returns (uint256 payout_) {
```
No more redeem all. Can make this external.

JExtor commented on contracts/Treasury.sol:
```diff
@@ -103,7 +109,7 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
     /* ========== MUTATIVE FUNCTIONS ========== */
 
     /**
-     * @notice allow approved address to deposit an asset for OHM
```
For deposit, only accepts a LIQUIDITYTOKEN or RESERVETOKEN. There are 2 other types of tokens here (XTOKEN and RISKRESERVETOKEN) will XTOKEN also be added as LIQUIDITYTOKEN and RISKRESERVETOKEN also be added as RESERVETOKEN?

0xCaps replied:
Will add documentation to clarify the role of XTOKEN.

JExtor commented on contracts/Treasury.sol:
```diff
@@ -236,23 +259,23 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+        emit RepayDebt(msg.sender, address(FLOOR), _amount, _amount);
     }
 
     /* ========== MANAGERIAL FUNCTIONS ========== */
```
Same with `auditReserves()` as `deposit()`

JExtor replied:
Or just have another check to the registry with XTOKEN and RISKRESERVETOKENS

JExtor commented on contracts/vesting/Vesting.sol:
```diff
@@ -140,55 +155,44 @@ contract InvestorClaimV2 is Ownable {
     function circulatingSupply() public view returns (uint256) {
-        return treasury.baseSupply().sub(ohm.balanceOf(dao));
-    }
+        return treasury.baseSupply().sub(FLOOR.balanceOf(authority.governor()));
```
Who will be receiving bond fees the governor or guardian? May be same to start, but would assume it would be the guardian (DAO) address rather than a governance contract. Could be governance though, just want to check.

JExtor commented on contracts/vesting/Vesting.sol:
```diff
@@ -140,55 +155,44 @@ contract InvestorClaimV2 is Ownable {
 
     /**
-     * @notice view OHM claimable for address. DAI decimals (18).
+     * @notice view FLOOR claimable for address. WETH decimals (18).
```
Would return amount FLOOR can claim in FLOOR decimals (9) not wETH decimals (18)

JExtor commented on contracts/vesting/Vesting.sol:
```diff
@@ -39,79 +34,99 @@ contract InvestorClaimV2 is Ownable {
+        uint256 redeemableForValue = redeemableFor(msg.sender);
+
+        // Sense check the amount requested
+        require(_amount > 0, "Nothing to claim");
```
Not really needed. Good check to see if for some reason vester passed in 0. Same as 124 also.

JExtor commented on contracts/vesting/Vesting.sol:
```diff
@@ -39,79 +34,99 @@ contract InvestorClaimV2 is Ownable {
+
+        // Sense check the amount requested
+        require(_amount > 0, "Nothing to claim");
+        require(redeemableForValue.mul(1e6) >= _amount, "Claim more than vested");
```
Redundant. Essentially same as 125

0xCaps replied:
Will try dropping this to see the effect.

JExtor commented on contracts/Treasury.sol:
```diff
@@ -174,21 +180,38 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
 
     /**
-     * @notice mint new OHM using excess reserves
+     * @notice allocators can manage assets without being limited by excessReserves
```
Potentially dangerous not needing to keep that backing of 0.001 ETH

0xCaps replied:
Will add documentation to clarify that reserves must be re-added to the treasury immediately after they are managed.

JExtor commented on contracts/Treasury.sol:
```diff
@@ -174,21 +180,38 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
     /**
-     * @notice mint new OHM using excess reserves
+     * @notice allocators can manage assets without being limited by excessReserves
+     * @notice always ensure the reserves are repalaced
```
replaced*?

JExtor commented on contracts/Treasury.sol:
```diff
@@ -174,21 +180,38 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+
+        if (permissions[STATUS.RESERVETOKEN][_token] || permissions[STATUS.LIQUIDITYTOKEN][_token]) {
+            uint256 value = tokenValue(_token, _amount);
+            totalReserves = (value < totalReserves) ? totalReserves.sub(value) : 0;
```
Is there ever a situation where `totalReserves` is 0? I can see possibly `excessReserves()`, might be better having a conditional in `excessReserves()` that does the check and returns 0 if so. Would be careful though, because then can essentially mint with not enough reserves in the treasury to keep that 0.001 wETH per FLOOR.

JExtor commented on contracts/Treasury.sol:
```diff
@@ -300,23 +323,28 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
             if (!registered) {
                 registry[_status].push(_address);
 
-                if (_status == STATUS.LIQUIDITYTOKEN || _status == STATUS.RESERVETOKEN) {
```
Good catch. Would you want to do the check with XTOKEN and RISKRESERVETOKEN as well? Making sure 1 is not the other?

JExtor commented on contracts/Treasury.sol:
```diff
@@ -349,6 +377,17 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+     * @param _token address
+     * @param _valuation uint256
+     */
+    function setRiskOffValuation(address _token, uint256 _valuation) external onlyPolicy {
```
Makes sense, looks good.

JExtor commented on contracts/Treasury.sol:
```diff
@@ -456,21 +502,61 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+            return IBondingCalculator(bondCalculator[_token])
+              .valuation(_token, _amount)
+              .mul(10**IERC20Metadata(address(FLOOR)).decimals())
+              .div(10**IERC20Metadata(_token).decimals())
```
would just switch this and line 539, so all mul is first. Or could just do something like:
`.div(10**(IERC20Metadata(_token).decimals().sub(10**3)))`

JExtor commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,291 @@
+     * Deprecated in favour of harvestAll(address _token).
+     */
+
+    function harvest(address _token, uint256 _amount) external override {
```
Why have this function if just using `harvestAll()`

kphed replied:
+1

drahrealm replied:
If it is going to be kept, at its current state, you can also set the mutability to `pure`

JExtor commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,291 @@
+     * @notice accounting of deposit / withdrawal of assets.
+     */
+
+    function accountingFor(
```
If not actually using amount deposited to check, may not even needed this function.

0xCaps committed:
Merge pull request #9 from FloorDAO/fixup/manage-xtokens-before-harvestall

0xCaps committed:
Update Vesting.sol

0xCaps committed:
Update NFTXAllocator.sol

0xNom committed:
Update claim rewards flow

0xNom committed:
Permission updates

0xNom committed:
Update test

0xCaps committed:
fixup

0xCaps committed:
fixup

0xNom committed:
Update test to RESERVETOKEN

JExtor commented on contracts/calculators/NFTXXTokenWethCalculator.sol:
```diff
@@ -0,0 +1,44 @@
+    using FixedPoint for *;
+    using SafeMath for uint256;
+
+    INFTXLPStaking public nftxLpStaking;
```
Can make these imutable

JExtor commented on contracts/aFloorMigration.sol:
```diff
@@ -0,0 +1,66 @@
+
+    uint256 public swapEndBlock;
+
+    IERC20 public FLOOR;
```
Both can be immutable.

0xNom replied:
We set this in an initialize call and not the constructor unfortunately so can't set them to `immutable`

JExtor commented on contracts/Treasury.sol:
```diff
@@ -456,21 +502,61 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+     * @param _token address
+     * @return uint256
+     */
+    function riskOffValuation(address _token) public view override returns (uint256) {
```
Makes sense

JExtor commented on contracts/BondDepository.sol:
```diff
@@ -8,11 +8,11 @@ import "./libraries/SafeERC20.sol";
 /// Review by: JeffX
 
-contract OlympusBondDepositoryV2 is IBondDepository, NoteKeeper {
+contract FloorBondDepository is IBondDepository, NoteKeeper {
```
Using require statements, can use custom errors if wanted. https://blog.soliditylang.org/2021/04/21/custom-errors/

JExtor commented on contracts/aFloorERC20.sol:
```diff
@@ -0,0 +1,28 @@
+contract AlphaFLOOR is ERC20 {
+    using SafeMath for uint256;
+
+    address public minter;
```
Can be immutable

JExtor commented on contracts/aFloorMigration.sol:
```diff
@@ -0,0 +1,66 @@
+contract AlphaFloorMigration is Ownable {
+    using SafeMath for uint256;
+
+    uint256 public swapEndBlock;
```
Why have a swap end block?

JExtor commented on contracts/aFloorMigration.sol:
```diff
@@ -0,0 +1,66 @@
+        );
+        require(block.number < swapEndBlock, "swapping of aFLOOR has ended");
+
+        aFLOOR.transferFrom(msg.sender, address(this), _amount);
```
Would import SafeERC20 and use safe transfers.

JExtor commented on contracts/aFloorMigration.sol:
```diff
@@ -0,0 +1,66 @@
+    /**
+      * @notice governor can withdraw any remaining FLOOR after swapEndBlock
+      */
+    function withdraw() external onlyOwner() {
```
Okay, so have the end block so owner can withdraw. I personally don't know if should be a time limit to allow users to redeem their aFLOOR

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+        // Deposit the xToken back into the treasury, increasing total reserves and minting 0 FLOOR
+        uint256 xTokenValue = treasury.tokenValue(dividendTokenInfo[_token].xToken, xTokenBalance);
+
+        IERC20(dividendTokenInfo[_token].xToken).approve(address(treasury), xTokenBalance);
```
Should use `safeApprove`

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+        uint256 value = treasury.tokenValue(_rewardToken, balance);
+
+        // Approve and deposit asset into treasury
+        IERC20(_rewardToken).approve(address(treasury), balance);
```
Should use `safeApprove`

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+     */
+
+    function harvestAll(address _token) override external {
+        // We only want to allow harvesting from a specified liquidity pool mapping
```
Variables may be a cleaner alternative to repeatedly accessing the mapping w/ bracket notation.

```
stakingTokenData memory s = stakingTokenInfo[_token];
dividendTokenData memory d = dividendTokenInfo[_token];
```

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+     * @notice claims rewards from the vault.
+     */
+
+    function harvestAll(address _token) override external {
```
Should switch `external` and `override` ordering for consistency

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+     */
+
+    function deposit(address _token, uint256 _amount) external override onlyPolicy {
+        require(stakingTokenInfo[_token].exists, "Unsupported staking token");
```
Variables may be a cleaner alternative to repeatedly accessing the mapping w/ bracket notation.

```
stakingTokenData memory s = stakingTokenInfo[_token];
dividendTokenData memory d = dividendTokenInfo[_token];
```

kphed replied:
There are other instances below but I think you get the point.

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.7.5;
+
+import "../libraries/Address.sol";
```
Unused import

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+
+        // Approve and deposit into inventory pool, returning xToken
+        if (stakingTokenInfo[_token].isLiquidityPool) {
+            IERC20(_token).approve(address(liquidityStaking), _amount);
```
Should use `safeApprove`

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+            IERC20(_token).approve(address(liquidityStaking), _amount);
+            liquidityStaking.deposit(stakingTokenInfo[_token].vaultId, _amount);
+        } else {
+            IERC20(_token).approve(address(inventoryStaking), _amount);
```
Should use `safeApprove`

kphed replied:
There are other instances below but I think you get the point.

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+     */
+
+    function setStakingToken(address _token, address _rewardToken, uint256 _vaultId, bool _isLiquidityPool) external override onlyPolicy {
+        require(_token != address(0), "Cannot set vault for NULL token");
```
Should `_rewardToken` and `_vaultId` be checked?

kphed replied:
Also, if you don't want staking tokens to be overwritten, should check the `exists` field like the `addDividendToken` method does

0xCaps replied:
Will also add `removeDividendToken()`

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+        require(_token != address(0), "Cannot set vault for NULL token");
+
+        // Set up our vault mapping information
+        stakingTokenInfo[_token].vaultId = _vaultId;
```
For consistency's sake, may want to assign to an initialized struct - which is what's being done in `addDividendToken`

kphed commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,303 @@
+     * @notice remove vault mapping.
+     */
+
+    function removeStakingToken(address _token) external override onlyPolicy {
```
Would a similar method be necessary for removing dividend tokens?

kphed commented on test/floor/allocators/NFTXAllocator.js:
```diff
@@ -0,0 +1,647 @@
+    expect(allocator.harvestAll(TEST_ADDRESS_1)).to.be.revertedWith('Must be liquidity staking token');
+
+    // Trigger our rewards to be claimed against the mapped vault ID
+    await allocator.harvestAll(TEST_ADDRESS_2);
```
Getting this error

`Error: Transaction reverted: function call to a non-contract account`

kphed commented on contracts/aFloorMigration.sol:
```diff
@@ -0,0 +1,66 @@
+        uint256 _swapDuration
+    ) public notInitialized() onlyOwner() {
+        require(_swapDuration < 2_000_000, "swap duration too long");
+        FLOOR = IERC20(_FLOOR);
```
Might be a good idea to check for zero addresses since there's no changing the variables after they're set.

kphed commented on contracts/aFloorMigration.sol:
```diff
@@ -0,0 +1,66 @@
+        _;
+    }
+
+    constructor() {}
```
Should be removed

0xNom replied:
Snippet is now outdated; We can no longer remove this constructor as it initialises the Floor Authority logic.

```
    constructor(
        address _authority
    ) FloorAccessControlled(IFloorAuthority(_authority)) {}
```

kphed reviewed (commented):
Hope this helps!

0xNom committed:
Add ERC721 support to treasury

0xNom committed:
Sync treasury mock

0xNom committed:
Update test to remove approval check

0xCaps committed:
Updating rinkeby/mainnet deploy scripts

0xCaps committed:
switched ownable to FloorAuthority

0xCaps committed:
Merge pull request #12 from FloorDAO/feat/deploy-script-update

0xCaps committed:
implement ERC721Receiver for safeTransferFrom

0xCaps committed:
Merge pull request #11 from FloorDAO/feature/treasury-erc721

0xCaps committed:
Merge pull request #10 from FloorDAO/feature/treasury-claim-rewards

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * @notice remove vault mapping.
+     */
+
+    function removeStakingToken(address _token) external override onlyPolicy {
```
change to onlyGovernor

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * @notice set vault mapping.
+     */
+
+    function setStakingToken(address _token, address _rewardToken, uint256 _vaultId, bool _isLiquidityPool) external override onlyPolicy {
```
change to onlyGovernor

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * @notice adds asset and corresponding xToken to mapping
+     */
+
+    function addDividendToken(address _token, address _xToken) external override onlyPolicy {
```
change to onlyGovernor

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * in the same transaction as `deposit()` because of a 2 second timelock in NFTX.
+     */
+
+    function depositXTokenToTreasury(address _token) external onlyPolicy {
```
change to onlyGovernor

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * @notice Withdraws from staking pool, and deposits asset into treasury.
+     */
+
+    function withdraw(address _token, uint256 _amount) external override onlyPolicy {
```
change to onlyGovernor

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * @notice withdraws asset from treasury, deposits asset into NFTX staking.
+     */
+
+    function deposit(address _token, uint256 _amount) external override onlyPolicy {
```
change to onlyGovernor

0xCaps commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     * @notice claims rewards from the vault.
+     */
+
+    function harvestAll(address _token) override external onlyPolicy {
```
change to onlyGovernor

0xCaps replied:
If rewards are earned while the xPUNKWETH token is held in the Allocator then the Allocator will have rewards to claim.

Wondering if we re-instate the old `harvestAll()` and rename this method to `treasuryHarvestAll()` ?

drahrealm commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+     */
+
+    function withdraw(address _token, uint256 _amount) external override onlyPolicy {
+        require(stakingTokenInfo[_token].exists, "Unsupported staking token");
```
Since `stakingTokenInfo` is being accessed a lot as a storage variable, might be a good idea to save some gas by storing in a temp variable first before using it

drahrealm commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+
+    function harvestAll(address _token) override external onlyPolicy {
+        // We only want to allow harvesting from a specified liquidity pool mapping
+        require(stakingTokenInfo[_token].exists, "Unsupported token");
```
Idem with my comment below regarding saving some gas

drahrealm commented on contracts/calculators/TokenWethCalculator.sol:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    function valuation(address _pair, uint256 amount_) external view override returns (uint256 _value) {
+        (uint256 reserve0, uint256 reserve1, ) = IUniswapV2Pair(_pair).getReserves();
```
You can consider storing the `IUniswapV2Pair(_pair)` into a variable before accessing it since you are calling it quite often

drahrealm commented on contracts/allocators/NFTXAllocator.sol:
```diff
@@ -0,0 +1,254 @@
+
+    function depositXTokenToTreasury(address _token) external onlyPolicy {
+        require(stakingTokenInfo[_token].exists, "Unsupported staking token");
+        require(dividendTokenInfo[_token].underlying == _token, "Unsupported dividend token");
```
Can save some gas too for `dividendTokenInfo`

0xNom committed:
Updates against PR

0xCaps commented on contracts/Treasury.sol:
```diff
@@ -43,9 +45,12 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
-        OHMDEBTOR
+        SFLOOR,
+        FLOORDEBTOR,
+        XTOKEN, // Any token that requires a calculator to determine its value
```
Updated to:

`// Any RESERVETOKEN that requires a calculator to determine its value`

0xCaps commented on contracts/Treasury.sol:
```diff
@@ -43,9 +45,12 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+        SFLOOR,
+        FLOORDEBTOR,
+        XTOKEN, // Any token that requires a calculator to determine its value
+        ALLOCATOR
```
`// Grants permission to withdraw Treasury reserves`

0xCaps commented on contracts/Treasury.sol:
```diff
@@ -43,9 +45,12 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
         LIQUIDITYTOKEN,
         LIQUIDITYMANAGER,
         RESERVEDEBTOR,
+        RISKRESERVETOKEN,
```
`// Allows a RESERVETOKEN to use a custom riskOffValuation when calculating tokenValue`

0xCaps commented on contracts/vesting/Vesting.sol:
```diff
@@ -39,79 +34,99 @@ contract InvestorClaimV2 is Ownable {
-        bool _rebasing,
-        bool _claimFromStaking
-    ) external {
+    function stake(address _to, uint256 _amount, bool _rebasing, bool _claimFromStaking) external {
```
The claim `_amount` is not multiplied by `1e6`

As it won't be used on the front end and not an important method, we can remove this method entirely.

0xCaps commented on contracts/Treasury.sol:
```diff
@@ -174,21 +184,95 @@ contract OlympusTreasury is OlympusAccessControlled, ITreasury {
+            return;
+        }
+
+        uint256 balanceDifference = previousBalance - newBalance;
```
This should be updated to `uint256 balanceDifference = newBalance.sub(previousBalance);`

0xNom committed:
PR updates

0xNom committed:
Validate FLOOR value without dust loss

0xNom committed:
Remove unused address import

0xNom committed:
Add rescueRewards method on allocator

0xCaps committed:
Merge pull request #13 from FloorDAO/feature/pr-review-amends

0xCaps committed:
deploy script updates

0xCaps committed:
deploy script updates
